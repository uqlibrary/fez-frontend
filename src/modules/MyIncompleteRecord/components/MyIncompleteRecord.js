/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */
/* eslint-disable react/jsx-no-undef */
/* eslint-disable no-trailing-spaces */
/* eslint-disable no-unreachable */
/* eslint-disable no-debugger */
/* eslint-disable no-constant-condition */
/* eslint-disable react-hooks/exhaustive-deps */
/* eslint-disable max-len */
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

import { ConfirmationBox } from 'modules/SharedComponents/Toolbox/ConfirmDialogBox';
import { ConfirmDiscardFormChanges } from 'modules/SharedComponents/ConfirmDiscardFormChanges';
import { useConfirmationState, userIsAdmin, useValidatedForm } from 'hooks';

import Button from '@mui/material/Button';
import Grid from '@mui/material/Unstable_Grid2';
import Typography from '@mui/material/Typography';

import { TextField } from 'modules/SharedComponents/Toolbox/TextField';
import { StandardPage } from 'modules/SharedComponents/Toolbox/StandardPage';
import { NavigationDialogBox } from 'modules/SharedComponents/Toolbox/NavigationPrompt';
import { StandardCard } from 'modules/SharedComponents/Toolbox/StandardCard';
import { Alert } from 'modules/SharedComponents/Toolbox/Alert';
import { FileUploadField } from 'modules/SharedComponents/Toolbox/FileUploader';
import { PublicationCitation } from 'modules/SharedComponents/PublicationCitation';
import { ContributorsEditorField } from 'modules/SharedComponents/ContributorsEditor';
import { NtroFields } from 'modules/SharedComponents/Toolbox/NtroFields';

import { validation, pathConfig, incompleteRecord } from 'config';
import { default as pagesLocale } from '../locale';
import { default as formsLocale } from 'locale/forms';
import { default as viewRecordLocale } from 'locale/viewRecord';
import { default as alertLocale } from 'locale/publicationForm';
import { useNavigate, useParams } from 'react-router-dom';
import { authorAffiliationRequired } from '../../../config/validation';
import { locale } from '../../../locale';
import {
    AFFILIATION_TYPE_NOT_UQ,
    AFFILIATION_TYPE_UQ,
    CPEE_NTRO_SUBTYPES,
    DOCUMENT_TYPE_BOOK_CHAPTER,
    DOCUMENT_TYPE_JOURNAL_ARTICLE,
    LP_NTRO_SUBTYPES,
    NTRO_SUBTYPES,
    ORG_TYPE_NOT_SET,
} from '../../../config/general';
import { leftJoin } from '../../../helpers/general';
import { isAdded } from '../../../helpers/datastreams';
import { useDispatch, useSelector } from 'react-redux';
import { clearFixRecord, loadRecordToFix } from '../../../actions';
import { InlineLoader } from '../../SharedComponents/Toolbox/Loaders';
import * as actions from 'actions/incompleteRecords';
import Field from '../../SharedComponents/Toolbox/ReactHookForm/components/Field';
import { FormProvider, useWatch } from 'react-hook-form';

export const FORM_NAME = 'MyIncompleteRecord';

const ntroFieldsDefaultValues = {
    significance: '',
    impactStatement: '',
    ntroAbstract: '',
    fez_record_search_key_ismn: '',
    fez_record_search_key_isrc: '',
    fez_record_search_key_series: {
        rek_series: '',
    },
    fez_record_search_key_volume_number: {
        rek_volume_number: '',
    },
    fez_record_search_key_issue_number: {
        rek_issue_number: '',
    },
    fez_record_search_key_start_page: {
        rek_start_page: '',
    },
    fez_record_search_key_end_page: {
        rek_end_page: '',
    },
    fez_record_search_key_total_pages: {
        rek_total_pages: '',
    },
    fez_record_search_key_original_format: {
        rek_original_format: '',
    },
    fez_record_search_key_audience_size: {
        rek_audience_size: '',
    },
    languages: '',
    qualityIndicators: '',
    grants: '',
};

const defaultValues = {
    ...ntroFieldsDefaultValues,
    authorsAffiliation: '',
    comments: '',
    files: '',
};

const getInitialValues = (recordToFix, author, isAdmin, disableInitialGrants) => {
    if (!recordToFix) {
        return defaultValues;
    }

    const grants = recordToFix.fez_record_search_key_grant_agency?.map((grantAgency, index) => ({
        grantAgencyName: grantAgency.rek_grant_agency,
        grantId:
            (recordToFix.fez_record_search_key_grant_id &&
                recordToFix.fez_record_search_key_grant_id.length > 0 &&
                recordToFix.fez_record_search_key_grant_id[index] &&
                recordToFix.fez_record_search_key_grant_id[index].rek_grant_id) ||
            '',
        grantAgencyType:
            (recordToFix.fez_record_search_key_grant_agency_type &&
                recordToFix.fez_record_search_key_grant_agency_type.length > 0 &&
                recordToFix.fez_record_search_key_grant_agency_type[index] &&
                recordToFix.fez_record_search_key_grant_agency_type[index].rek_grant_agency_type) ||
            ORG_TYPE_NOT_SET,
        disabled: disableInitialGrants,
    }));

    const affiliationDataMap = [
        {
            infoArray: recordToFix.fez_record_search_key_author,
            key: 'rek_author_order',
        },
        {
            infoArray: recordToFix.fez_record_search_key_author_affiliation_name,
            key: 'rek_author_affiliation_name_order',
        },
        {
            infoArray: recordToFix.fez_record_search_key_author_affiliation_type,
            key: 'rek_author_affiliation_type_order',
        },
    ];

    const authorsAffiliation = affiliationDataMap
        .reduce(
            (authors, affiliationData) =>
                leftJoin(authors, affiliationData.infoArray, 'rek_author_id_order', affiliationData.key),
            recordToFix.fez_record_search_key_author_id,
        )
        .map(authorAffiliation => ({
            affiliation:
                authorAffiliation.rek_author_affiliation_name === locale.global.orgTitle
                    ? AFFILIATION_TYPE_UQ
                    : AFFILIATION_TYPE_NOT_UQ,
            creatorRole: '',
            nameAsPublished: authorAffiliation.rek_author,
            orgaff: authorAffiliation.rek_author_affiliation_name || '',
            orgtype:
                (authorAffiliation.rek_author_affiliation_type &&
                    String(authorAffiliation.rek_author_affiliation_type)) ||
                '',
            uqIdentifier: String(authorAffiliation.rek_author_id),
            disabled: authorAffiliation.rek_author_id && authorAffiliation.rek_author_id !== author.aut_id,
        }))
        .map(authorAffiliation => ({
            ...authorAffiliation,
            required: authorAffiliationRequired(authorAffiliation, author),
        }));

    const significance = isAdmin && recordToFix.fez_record_search_key_significance;

    const languages = (recordToFix &&
        (recordToFix.fez_record_search_key_language || []).length > 0 &&
        recordToFix.fez_record_search_key_language.map(lang => lang.rek_language)) || ['eng'];

    return {
        ...defaultValues,
        authorsAffiliation,
        grants,
        languages,
        significance,
    };
};

const getCurrentAuthorOrder = (recordToFix, author) => {
    const currentAuthor =
        recordToFix &&
        recordToFix.fez_record_search_key_author_id.filter(authorId => authorId.rek_author_id === author.aut_id);
    return !!currentAuthor && currentAuthor.length > 0 && currentAuthor[0].rek_author_id_order;
};

const getNtroFieldFlags = (recordToFix, author) => {
    const currentAuthorOrder = getCurrentAuthorOrder(recordToFix, author);
    const significance = (recordToFix.fez_record_search_key_significance || []).filter(
        item => item.rek_significance_order === currentAuthorOrder,
    );
    const contributionStatement = (recordToFix.fez_record_search_key_creator_contribution_statement || []).filter(
        item => item.rek_creator_contribution_statement_order === currentAuthorOrder,
    );

    return {
        hideAbstract: !!recordToFix.rek_formatted_abstract || !!recordToFix.rek_description,
        hideLanguage: (recordToFix.fez_record_search_key_language || []).length !== 0,
        hidePeerReviewActivity: (recordToFix.fez_record_search_key_quality_indicator || []).length !== 0,
        hideExtent:
            [DOCUMENT_TYPE_BOOK_CHAPTER, DOCUMENT_TYPE_JOURNAL_ARTICLE].includes(recordToFix.rek_display_type_lookup) ||
            !!(recordToFix.fez_record_search_key_total_pages || {}).rek_total_pages,
        hideAudienceSize:
            ![...LP_NTRO_SUBTYPES, ...CPEE_NTRO_SUBTYPES].includes(recordToFix.rek_subtype) ||
            !!(recordToFix.fez_record_search_key_audience_size || {}).rek_audience_size,
        showSignificance: significance.length === 0 || (significance.length > 0 && !significance[0].rek_significance),
        showContributionStatement:
            contributionStatement.length === 0 ||
            (contributionStatement.length > 0 &&
                (!contributionStatement[0].rek_creator_contribution_statement ||
                    contributionStatement[0].rek_creator_contribution_statement === '' ||
                    contributionStatement[0].rek_creator_contribution_statement ===
                        locale.global.defaultAuthorDataPlaceholder)),
    };
};

const isLoggedInUserLinked = (author, recordToFix, searchKey, subkey) => {
    return (
        !!author &&
        !!recordToFix &&
        recordToFix[searchKey] &&
        recordToFix[searchKey].length > 0 &&
        recordToFix[searchKey].some(authorId => authorId[subkey] === author.aut_id)
    );
};

const getIsAuthorLinked = (recordToFix, author) => {
    const isAuthorLinked = isLoggedInUserLinked(
        author,
        recordToFix,
        'fez_record_search_key_author_id',
        'rek_author_id',
    );
    const isContributorLinked = isLoggedInUserLinked(
        author,
        recordToFix,
        'fez_record_search_key_contributor_id',
        'rek_contributor_id',
    );

    return isAuthorLinked || isContributorLinked;
};

export const formLevelValidation = (data, author) => {
    const { authorsAffiliation } = data;

    const formErrors = {};
    if (authorsAffiliation?.some?.(authorAffiliation => authorAffiliationRequired(authorAffiliation, author))) {
        formErrors.authorsAffiliation = locale.validationErrors.authorsAffiliationIncomplete;
    }
    return formErrors;
};

const isFileValid = dataStream => {
    const {
        files: { blacklist },
    } = incompleteRecord;
    return !dataStream.dsi_dsid.match(blacklist.namePrefixRegex) && isAdded(dataStream);
};

const MyIncompleteRecord = ({ disableDeleteAllGrants, disableInitialGrants }) => {
    const navigate = useNavigate();
    const dispatch = useDispatch();
    const [isOpen, showConfirmation, hideConfirmation] = useConfirmationState();
    const isAdmin = userIsAdmin();
    const { pid } = useParams();
    // constants
    const txt = pagesLocale;
    const txtFixForm = formsLocale.forms.fixPublicationForm;
    const authors = txt.fields.authors;
    // global state
    const { recordToFix, loadingRecordToFix } = useSelector(state => state.get('fixRecordReducer'));
    const { author, accountAuthorLoading } = useSelector(state => state.get('accountReducer'));
    // local state
    const [ntroFieldProps, setNtroFieldProps] = React.useState({});
    const [isNtro, setIsNtro] = React.useState(false);
    const [hasAnyFiles, setHasAnyFiles] = React.useState(false);
    const [isAuthorLinked, setIsAuthorLinked] = React.useState(true);
    const values = getInitialValues(recordToFix, author, isAdmin, disableInitialGrants);
    // form
    const form = useValidatedForm({ values });
    const {
        control,
        getPropsForAlert,
        safelyHandleSubmit,
        mergeWithFormValues,
        formState: { isDirty, isSubmitting, isSubmitSuccessful, hasValidationError },
    } = form;
    const [authorsAffiliation] = useWatch({ name: ['authorsAffiliation'], control });

    // load record on rendering
    useEffect(() => {
        pid && !recordToFix && !!loadRecordToFix && dispatch(loadRecordToFix(pid));
        return () => dispatch(clearFixRecord());
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [pid]);

    // update state upon loading recordToFix
    useEffect(() => {
        if (!!recordToFix) {
            setNtroFieldProps(getNtroFieldFlags(recordToFix, author));
            setIsNtro(!!recordToFix.rek_subtype && !!NTRO_SUBTYPES.includes(recordToFix.rek_subtype));
            setHasAnyFiles(recordToFix.fez_datastream_info.filter(isFileValid).length > 0);
            setIsAuthorLinked(getIsAuthorLinked(recordToFix, author));
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [recordToFix]);

    // show confirmation dialog upon successful form submission
    useEffect(() => {
        if (isSubmitSuccessful) showConfirmation();
    }, [showConfirmation, isSubmitSuccessful]);

    const fieldWrapper = React.useCallback(props => <Field {...props} control={control} />, [control]);

    // loading message
    if (accountAuthorLoading || loadingRecordToFix) {
        return <InlineLoader message={txt.loadingMessage} />;
    }

    // if author is not linked to this record, abandon form
    if (!isAuthorLinked) {
        navigate(-1);
        return <div id="author-not-linked" data-testid="author-not-linked" />;
    }

    const alertProps = validation.getErrorAlertProps({
        alertLocale: {
            ...alertLocale,
            progressAlert: txt.progressAlert,
            successAlert: txt.successAlert,
        },
        ...getPropsForAlert(formLevelValidation({ authorsAffiliation }, author)),
    });

    const navigateToMyIncomplete = () => {
        navigate(pathConfig.records.incomplete);
    };

    const navigateToDashboard = () => {
        navigate(pathConfig.dashboard);
    };

    const onSubmit = safelyHandleSubmit(async () => {
        const data = mergeWithFormValues({
            publication: { ...recordToFix },
            author: { ...author },
        });
        dispatch(await actions.updateIncompleteRecord(data));
    });

    return (
        <StandardPage title={txt.title} help={txt.help}>
            <PublicationCitation publication={recordToFix} hideContentIndicators citationStyle={'list'} />
            <ConfirmDiscardFormChanges dirty={isDirty} submitSucceeded={isSubmitSuccessful}>
                <FormProvider {...form}>
                    <form onSubmit={onSubmit}>
                        <NavigationDialogBox
                            when={isDirty && !isSubmitSuccessful}
                            txt={txtFixForm.cancelWorkflowConfirmation}
                        />
                        <ConfirmationBox
                            confirmationBoxId="submit-succeeded"
                            onAction={navigateToDashboard}
                            onClose={hideConfirmation}
                            onCancelAction={navigateToMyIncomplete}
                            isOpen={isOpen}
                            locale={txt.successWorkflowConfirmation}
                        />
                        <Grid container spacing={3}>
                            <Grid xs={12}>
                                <Alert title={txt.prompt.title} message={txt.prompt.message} type={txt.prompt.type} />
                            </Grid>
                            <Grid xs={12}>
                                <StandardCard title={viewRecordLocale.viewRecord.sections.publicationDetails}>
                                    <Grid
                                        container
                                        sx={{
                                            paddingBottom: '12px',
                                            borderBottom: '1px solid',
                                            borderBottomColor: 'secondary.light',
                                        }}
                                    >
                                        {!!recordToFix && !!recordToFix.rek_display_type_lookup && (
                                            <Grid container alignItems="flex-start" width={'100%'}>
                                                <Grid xs={12} sm={3}>
                                                    <Typography>
                                                        {
                                                            viewRecordLocale.viewRecord.headings.default
                                                                .publicationDetails.rek_display_type
                                                        }
                                                    </Typography>
                                                </Grid>
                                                <Grid xs={12} sm={9}>
                                                    <Typography>{recordToFix.rek_display_type_lookup}</Typography>
                                                </Grid>
                                            </Grid>
                                        )}
                                    </Grid>
                                    <Grid
                                        container
                                        sx={{
                                            marginTop: '12px',
                                            paddingBottom: '12px',
                                            borderBottom: '1px solid',
                                            borderBottomColor: 'secondary.light',
                                        }}
                                    >
                                        {!!recordToFix && !!recordToFix.rek_subtype && (
                                            <Grid container alignItems="flex-start" width={'100%'}>
                                                <Grid xs={12} sm={3}>
                                                    <Typography>
                                                        {
                                                            viewRecordLocale.viewRecord.headings.default
                                                                .publicationDetails.rek_subtype
                                                        }
                                                    </Typography>
                                                </Grid>
                                                <Grid xs={12} sm={9}>
                                                    <Typography>{recordToFix.rek_subtype}</Typography>
                                                </Grid>
                                            </Grid>
                                        )}
                                    </Grid>
                                </StandardCard>
                            </Grid>
                            {isNtro && (
                                <NtroFields
                                    fieldWrapper={fieldWrapper}
                                    submitting={isSubmitting}
                                    hideIsmn
                                    hideIsrc
                                    hideVolume
                                    hideIssue
                                    hideStartPage
                                    hideEndPage
                                    hideOriginalFormat
                                    hideSeries
                                    disableDeleteAllGrants={disableDeleteAllGrants}
                                    {...ntroFieldProps}
                                />
                            )}
                            <Grid xs={12}>
                                <StandardCard title={authors.title} help={authors.help}>
                                    <Typography>{authors.description}</Typography>
                                    <Field
                                        control={control}
                                        component={ContributorsEditorField}
                                        contributorEditorId="rek-author"
                                        editMode
                                        canEdit
                                        hideDelete
                                        hideReorder
                                        isNtro
                                        locale={txt.fields.authors.field}
                                        name="authorsAffiliation"
                                        required
                                        showContributorAssignment
                                    />
                                </StandardCard>
                            </Grid>

                            <Grid xs={12}>
                                <StandardCard title={txt.fields.notes.title}>
                                    <Field
                                        control={control}
                                        component={TextField}
                                        disabled={isSubmitting}
                                        fullWidth
                                        label={txt.fields.notes.label}
                                        multiline
                                        name="comments"
                                        placeholder={txt.fields.notes.placeholder}
                                        rows={5}
                                        style={{ marginTop: -24 }}
                                        textFieldId="comments"
                                        type="text"
                                    />
                                </StandardCard>
                            </Grid>
                            {!hasAnyFiles && (
                                <Grid xs={12}>
                                    <StandardCard title={txt.fields.fileUpload.title}>
                                        <Field
                                            control={control}
                                            name="files"
                                            component={FileUploadField}
                                            disabled={isSubmitting}
                                            requireOpenAccessStatus
                                            validate={[validation.fileUploadRequired, validation.validFileUpload]}
                                            isNtro
                                            {...txt.fields.fileUpload}
                                        />
                                    </StandardCard>
                                </Grid>
                            )}
                            {alertProps && (
                                <Grid xs={12}>
                                    <Alert pushToTop {...alertProps} />
                                </Grid>
                            )}
                        </Grid>
                        <Grid container spacing={3}>
                            <Grid xs sx={{ display: { xs: 'none', md: 'block' } }} />

                            <Grid xs={12} md="auto">
                                <Button
                                    id="cancel-fix-work"
                                    variant="contained"
                                    fullWidth
                                    children={txt.cancelButtonLabel}
                                    disabled={isSubmitting}
                                    onClick={navigateToMyIncomplete}
                                />
                            </Grid>
                            <Grid xs={12} md="auto">
                                <Button
                                    id="update-my-work"
                                    variant="contained"
                                    color="primary"
                                    fullWidth
                                    children={txt.submitButtonLabel}
                                    type="submit"
                                    disabled={isSubmitting || hasValidationError}
                                />
                            </Grid>
                        </Grid>
                    </form>
                </FormProvider>
            </ConfirmDiscardFormChanges>
        </StandardPage>
    );
};

MyIncompleteRecord.propTypes = {
    disableDeleteAllGrants: PropTypes.bool,
    disableInitialGrants: PropTypes.bool,
};

export default React.memo(MyIncompleteRecord);
