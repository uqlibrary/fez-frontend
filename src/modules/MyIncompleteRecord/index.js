export { default as MyIncompleteRecord } from './components/MyIncompleteRecord';
