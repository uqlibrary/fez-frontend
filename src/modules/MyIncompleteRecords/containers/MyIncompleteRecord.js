import { connect } from 'react-redux';
import { getFormSyncErrors } from 'redux-form/immutable';
import Immutable from 'immutable';
import { loadRecordToFix, clearFixRecord } from 'actions';

import { FORM_NAME } from '../components/MyIncompleteRecordForm';
import MyIncompleteRecordContainer from '../components/MyIncompleteRecordContainer';

const mapStateToProps = state => {
    const formErrors = getFormSyncErrors(FORM_NAME)(state) || Immutable.Map({});

    return {
        ...state.get('fixRecordReducer'),
        ...state.get('accountReducer'),
        formErrors: formErrors,
        disableSubmit: formErrors && !(formErrors instanceof Immutable.Map),
    };
};

const mapDispatchToProps = dispatch => ({
    loadRecordToFix: pid => dispatch(loadRecordToFix(pid)),
    clearFixRecord: () => dispatch(clearFixRecord()),
});

export default connect(mapStateToProps, mapDispatchToProps)(MyIncompleteRecordContainer);
