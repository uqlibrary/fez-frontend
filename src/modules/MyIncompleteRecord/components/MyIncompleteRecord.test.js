import React from 'react';
import { default as locale } from 'locale/global';
import { default as pageLocale } from '../locale';
import { default as alertLocale } from 'locale/publicationForm';
import {
    render,
    WithReduxStore,
    WithRouter,
    screen,
    assertEnabled,
    userEvent,
    mockUseValidatedForm,
    waitForTextToBeRemoved,
    waitForText,
    preview,
    waitToBeDisabled,
    waitToBeEnabled,
} from 'test-utils';
import Immutable from 'immutable';
import { waitFor } from '@testing-library/dom';

jest.mock('../../../context');

import MyIncompleteRecord from './MyIncompleteRecord';
import { incompleteNTRORecordUQ352045 } from '../../../mock/data';
import validationErrors from '../../../locale/validationErrors';
import * as repositories from '../../../repositories';
import { pathConfig } from '../../../config';

const mockUseNavigate = jest.fn();
let mockParams;
/* eslint-disable react/prop-types */
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate,
    useParams: () => mockParams,
}));

function setup(props = {}) {
    props.author = props.hasOwnProperty('author') ? props.author : { aut_id: 79324 };
    const state = Immutable.Map({
        fixRecordReducer: {
            recordToFix: props.publication,
            loadingRecordToFix: props.hasOwnProperty('loadingRecordToFix')
                ? props.loadingRecordToFix
                : !props.publication,
        },
        accountReducer: {
            author: props.author,
            accountAuthorLoading: props.hasOwnProperty('accountAuthorLoading')
                ? props.accountAuthorLoading
                : !props.author,
        },
    });

    return render(
        <WithReduxStore initialState={state}>
            <WithRouter>
                <MyIncompleteRecord
                    disableDeleteAllGrants={props.disableDeleteAllGrants}
                    disableInitialGrants={props.disableInitialGrants}
                />
            </WithRouter>
        </WithReduxStore>,
    );
}

describe('MyIncompleteRecord', () => {
    const isDebugging = true;
    const waitForOptions = { timeout: isDebugging ? 120000 : 1000 };

    const mockRichEditorFieldValues = impactStatement =>
        mockUseValidatedForm((props, original) => {
            props.values.impactStatement = impactStatement !== undefined ? impactStatement : 'impact statement';
            return original(props);
        });

    const assertValidationErrorSummary = async () => {
        await waitForText(/Form cannot be submitted until all fields are valid/, waitForOptions);
        assertEnabled('incomplete-record-button-cancel');
        await waitToBeDisabled('incomplete-record-button-update');
    };

    const assertNoValidationErrorSummary = async () => {
        await waitForTextToBeRemoved(alertLocale.validationAlert.message, waitForOptions);
        assertEnabled('incomplete-record-button-cancel');
        await waitToBeEnabled('incomplete-record-button-update');
    };

    const fillUpForm = async ({ waitForFieldErrorToBeCleared, waitForValidationSummaryRemoval }) => {
        await userEvent.click(screen.getByTestId('rek-significance-select'));
        await userEvent.click(screen.queryByText('Major'));
        waitForFieldErrorToBeCleared && (await waitForTextToBeRemoved('Scale/Significance of work is required'));

        await userEvent.click(screen.getByTestId('rek-audience-size-select'));
        await userEvent.click(screen.queryByText('100 - 500'));
        waitForFieldErrorToBeCleared && (await waitForTextToBeRemoved('Audience size is required'));

        await userEvent.click(screen.getByTestId('rek-quality-indicator-select'));
        await userEvent.click(screen.queryByText('Disseminated via nationally recognised outlet or entity'));
        await userEvent.tab();
        waitForFieldErrorToBeCleared && (await waitForTextToBeRemoved('Quality indicator is required'));

        await userEvent.click(screen.getByTestId('rek-author-list-row-0-edit'));
        await userEvent.click(screen.getByTestId('org-affiliation-select'));
        await userEvent.click(screen.queryByText('UQ'));
        await userEvent.click(screen.getByRole('button', { name: 'Change Details' }));
        waitForFieldErrorToBeCleared &&
            (await waitForTextToBeRemoved('Author affiliation rows marked with red are required'));

        waitForValidationSummaryRemoval && (await assertNoValidationErrorSummary());
    };

    const submitForm = async () => await userEvent.click(screen.getByTestId('incomplete-record-button-update'));

    const retryUpload = async () => {
        await waitFor(
            () => screen.getByText(new RegExp(formLocale.thesisSubmission.fileUpload.failedAlertLocale.title, 'i')),
            waitForOptions,
        );
        await userEvent.click(
            screen.getByText(formLocale.thesisSubmission.fileUpload.failedAlertLocale.actionButtonLabel),
        );
    };

    let mockRecordToFix = incompleteNTRORecordUQ352045;
    beforeEach(() => {
        mockParams = { pid: mockRecordToFix.rek_pid };
        mockRecordToFix = {
            ...incompleteNTRORecordUQ352045,
            fez_record_search_key_author: [
                {
                    ...incompleteNTRORecordUQ352045.fez_record_search_key_author[1],
                    rek_author_order: 1,
                },
            ],
            fez_record_search_key_author_id: [
                {
                    ...incompleteNTRORecordUQ352045.fez_record_search_key_author_id[1],
                    rek_author_id_order: 1,
                },
            ],
        };

        jest.restoreAllMocks();
        mockSessionApi.resetHandlers();
        mockApi.resetHandlers();
    });

    afterEach(() => {
        jest.restoreAllMocks();
        mockSessionApi.resetHandlers();
        mockApi.resetHandlers();
    });

    it('should render 404 page for unexisting record', async () => {
        jest.spyOn(require('actions/fixRecord'), 'loadRecordToFix').mockImplementationOnce(() => ({
            type: 'FIX_RECORD_LOAD_FAILED',
        }));
        setup();
        await waitForText(/Work not found/i, waitForOptions);
    });

    it('should render loader when author is loading', () => {
        setup({ author: null });
        waitForText(pageLocale.loading, waitForOptions);
    });

    it('should render loader when record is loading', () => {
        setup();
        waitForText(pageLocale.loading, waitForOptions);
    });

    it('should redirect if author not linked', () => {
        setup({ publication: mockRecordToFix, author: { aut_id: 404 } });
        expect(mockUseNavigate).toHaveBeenCalledWith(-1);
    });

    it('should navigate to my incomplete records page', async () => {
        const { getByTestId } = setup({
            publication: {
                ...mockRecordToFix,
                fez_record_search_key_author_affiliation_name: [
                    {
                        rek_author_affiliation_name: locale.global.orgTitle,
                        rek_author_affiliation_name_order: 1,
                    },
                ],
                fez_record_search_key_author_affiliation_type: [
                    {
                        rek_author_affiliation_type: 453989,
                        rek_author_affiliation_type_order: 1,
                    },
                ],
                fez_record_search_key_language: [],
            },
        });
        await userEvent.click(getByTestId('incomplete-record-button-cancel'));

        expect(!!pathConfig.records.incomplete.length).toBeTruthy();
        expect(mockUseNavigate).toHaveBeenCalledWith(pathConfig.records.incomplete);
    });

    it('should display error summary according to invalid fields', async () => {
        setup({ publication: mockRecordToFix });
        await assertValidationErrorSummary();

        await waitForText('Scale/Significance of work is required');
        await waitForText('Creator research statement is required');
        await waitForText('Audience size is required');
        await waitForText('Quality indicator is required');
        await waitForText('Author affiliation rows marked with red are required');

        await fillUpForm({ waitForFieldErrorToBeCleared: true });
    });

    it('should display error summary according to invalid rich editor fields', async () => {
        mockRichEditorFieldValues();
        const { queryByText } = setup({ publication: mockRecordToFix });
        await assertValidationErrorSummary();

        expect(queryByText('Creator research statement is required')).not.toBeInTheDocument();
    });

    it('should display grant editor field error in error summary', async () => {
        const { getByTestId, getAllByText } = setup({ publication: mockRecordToFix });
        await assertValidationErrorSummary();

        await userEvent.type(getByTestId('rek-grant-agency-input'), 'grant');
        await waitFor(() => getAllByText(validationErrors.validationErrorsSummary.grants)[0], waitForOptions);
    });

    it('should navigate to dashboard after form submission', async () => {
        const pid = mockRecordToFix.rek_pid;
        mockApi
            .onPatch(repositories.routes.EXISTING_RECORD_API({ pid }).apiUrl)
            .replyOnce(200, { data: { rek_pid: pid } })
            .onPost(repositories.routes.RECORDS_ISSUES_API({ pid }).apiUrl)
            .replyOnce(200, { data: { pid } });
        // .onPost(repositories.routes.FILE_UPLOAD_API().apiUrl)
        // .replyOnce(200, 's3-ap-southeast-2.amazonaws.com')
        // .onPut('s3-ap-southeast-2.amazonaws.com')
        // .replyOnce(200, {});

        mockRichEditorFieldValues();
        const { getByTestId } = setup({ publication: mockRecordToFix });
        await assertValidationErrorSummary();
        await fillUpForm({ waitForValidationSummaryRemoval: true });
        await userEvent.type(getByTestId('comments-input'), 'comments');
        await submitForm();

        await waitForText(pageLocale.successWorkflowConfirmation.confirmationTitle, waitForOptions);
        await userEvent.click(getByTestId('confirm-submit-succeeded'));

        expect(!!pathConfig.dashboard.length).toBeTruthy();
        expect(mockUseNavigate).toHaveBeenCalledWith(pathConfig.dashboard);
    });
});
