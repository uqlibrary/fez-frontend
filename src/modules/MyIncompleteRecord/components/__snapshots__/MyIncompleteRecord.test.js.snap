// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyIncompleteRecord form submission payload all fields data 1`] = `"{"rek_pid":"UQ:352045","fez_record_search_key_audience_size":{"rek_audience_size":453993},"comments":"comments","fez_record_search_key_language":[{"rek_language":"eng","rek_language_order":1}],"fez_record_search_key_quality_indicator":[{"rek_quality_indicator":453996,"rek_quality_indicator_order":1}],"fez_record_search_key_significance":[{"rek_significance":454026,"rek_significance_order":1}],"fez_record_search_key_creator_contribution_statement":[{"rek_creator_contribution_statement":"<p>statement</p>","rek_creator_contribution_statement_order":1}],"fez_record_search_key_file_attachment_name":[{"rek_file_attachment_name":"myTestImage.png","rek_file_attachment_name_order":1}],"fez_record_search_key_file_attachment_embargo_date":[],"fez_record_search_key_file_attachment_access_condition":[{"rek_file_attachment_access_condition":1,"rek_file_attachment_access_condition_order":1}],"fez_record_search_key_file_attachment_security_policy":[],"fez_record_search_key_author_affiliation_name":[{"rek_author_affiliation_name":"The University of Queensland","rek_author_affiliation_name_order":1}],"fez_record_search_key_author_affiliation_type":[{"rek_author_affiliation_type":453989,"rek_author_affiliation_type_order":1}]}"`;

exports[`MyIncompleteRecord form submission payload all fields data 2`] = `"{"issue":"Added comments: \\n          comments \\n\\n                Added files: \\n          myTestImage.png"}"`;

exports[`MyIncompleteRecord form submission payload all fields data 3`] = `
File {
  "path": "myTestImage.png",
}
`;
